solr/core/src/test/org/apache/solr/util/TestObjectReleaseTracker.java

```diff
@@ -16,46 +16,103 @@
 
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.hamcrest.Matchers.containsString;
 
-@Limit(bytes=150000) // raise limit as this writes to sys err
 public class TestObjectReleaseTracker extends SolrTestCaseJ4 {
-  
+
   @Test
-  public void testObjectReleaseTracker() {
-    ObjectReleaseTracker.track(new Object());
-    ObjectReleaseTracker.release(new Object());
-    assertNotNull(SolrTestCaseJ4.clearObjectTrackerAndCheckEmpty(1));
-    assertNull(SolrTestCaseJ4.clearObjectTrackerAndCheckEmpty(1));
+  public void testReleaseObject() {
     Object obj = new Object();
     ObjectReleaseTracker.track(obj);
     ObjectReleaseTracker.release(obj);
     assertNull(SolrTestCaseJ4.clearObjectTrackerAndCheckEmpty(1));
-    
+
     Object obj1 = new Object();
     ObjectReleaseTracker.track(obj1);
     Object obj2 = new Object();
     ObjectReleaseTracker.track(obj2);
     Object obj3 = new Object();
     ObjectReleaseTracker.track(obj3);
-    
+
     ObjectReleaseTracker.release(obj1);
     ObjectReleaseTracker.release(obj2);
     ObjectReleaseTracker.release(obj3);
     assertNull(SolrTestCaseJ4.clearObjectTrackerAndCheckEmpty(1));
-    
+  }
+
+  @Test
+  public void testUnreleased() {
+    Object obj1 = new Object();
+    Object obj2 = new Object();
+    Object obj3 = new Object();
+
     ObjectReleaseTracker.track(obj1);
     ObjectReleaseTracker.track(obj2);
     ObjectReleaseTracker.track(obj3);
     
     ObjectReleaseTracker.release(obj1);
     ObjectReleaseTracker.release(obj2);
     // ObjectReleaseTracker.release(obj3);
+
     assertNotNull(SolrTestCaseJ4.clearObjectTrackerAndCheckEmpty(1));
     assertNull(SolrTestCaseJ4.clearObjectTrackerAndCheckEmpty(1));
   }
+
+  @Test
+  public void testReleaseDifferentObject() {
+    ObjectReleaseTracker.track(new Object());
+    ObjectReleaseTracker.release(new Object());
+    assertNotNull(SolrTestCaseJ4.clearObjectTrackerAndCheckEmpty(1));
+    assertNull(SolrTestCaseJ4.clearObjectTrackerAndCheckEmpty(1));
+  }
+
+  @Test
+  public void testAnonymousClasses() {
+    ObjectReleaseTracker.track(new Object() {});
+    String message = SolrTestCaseJ4.clearObjectTrackerAndCheckEmpty(1);
+    MatcherAssert.assertThat(message, containsString("[Object]"));
+  }
+
+  @Test
+  public void testAsyncTracking() throws InterruptedException, ExecutionException {
+    ExecutorService es = ExecutorUtil.newMDCAwareSingleThreadExecutor(new SolrNamedThreadFactory("TestExecutor"));
+    Object trackable = new Object();
+
+    Future<?> result = es.submit(() -> {
+      ObjectReleaseTracker.track(trackable);
+    });
+
+    result.get(); // make sure that track has been called
+    String message = SolrTestCaseJ4.clearObjectTrackerAndCheckEmpty(1);
+    MatcherAssert.assertThat(message, containsString(getTestName()));
```

Comment:
We could perhaps also test here that the async details are captured too. I'll add a small commit, feel free to revert or amend.

Reply:
Based on your commit, I realized I wasn't testing what I thought I was testing, so I tried a slightly different set of criteria.